crates/bevy_ecs/src/reflect/mod.rs

```diff
@@ -1,11 +1,58 @@
 //! Types that enable reflection support.
+//!
+//! # Automatic Reflect Type Registration
+//! The [`Component`] and [`Resource`] resource derive macros will automatically register
+//! its types that implement [`Reflect`] into the [`AppTypeRegistry`] resource when first
+//! seen by the ECS [`World`].
+//!
+//! If you find that your component or resource is not registered, they may need to be manually
+//! registered. There are a few exceptions:
+//!
+//!  * Automatic registration is only supported via the derive macros. Manual implementations of
+//!    `Component`, `Resource`, or `Reflect` must be manually registered.
+//!  * The associated ECS trait must be reflected (via `reflect(Component)` or `reflect(Resource)`).
```

Comment:
This was requested so that users could opt out of this, such as avoiding serialization in scenes. I personally think we should find other ways of opting out of each use case and just move towards registration by default.

Reply:
We could also consider an opt-out attribute. There could be other reasons to exclude a type from the registry besides opting out of scenes. For larger projects, you may only want to register types you know you’ll need (to save on memory and reduce startup times). You might also be using a third party crate that uses the registry for things you want to opt out of as well. Or maybe you don't want to "leak" type information to downstream crates.

Whatever the reason, we need a way to opt out of automatic registration since registration is intentionally a one-way process: you can't undo or remove a registration. This is so that guarantees about the registered type are never broken without the user being aware of it (unless, of course, a mischievous plugin goes nuclear and replaces the entire `AppTypeRegistry` resource but that's a whole separate issue haha).

Reply:
+1 for the opt-out attribute. It's more explicit, and allows this to work even when the user forgets to reflect `Component`/`Resource`